Fix random integer generation to match results from NumPy.

mpetroff opened the pull request:
This pull request modifies Numba's PRNG to match NumPy when using `np.random` (while not modifying the behavior of `random`), implementing much of what is described in #2782.

It adds support for 64-bit integers, which was missing in the patch in the aforementioned issue, and fixes a problem with power of two ranges (extending the unit tests to check for this). Support for CPython's PRNG initialization was not added.

seibert commented:
Looks like there is a failure on Windows, likely due to the fact that `int` is `int32` there, not `int64` like on Mac/Linux.  (I haven't traced through the code yet, but that is the top reason for things to fail on Windows...)

mpetroff commented:
It's odd, since the test seems to be failing in the non-JITed NumPy path.

stuartarchibald commented:
Thanks for the patch. Will review shortly. I think you just unluckily hit `np.random.RandomState.randint` having a return type of `np.int` opposed to there being anything fundamentally at fault. Also, if it fails CI due to utterly unrelated problems on Python 2.7 please try merging in master and that may well fix it!

stuartarchibald commented on numba/targets/randomimpl.py:
```diff
@@ -327,18 +335,41 @@ def _randrange_impl(context, builder, start, stop, step, state):
-    too_large = builder.icmp_signed('>=', r, n)
-    builder.cbranch(too_large, bbwhile, bbend)
+    if state == "np":
+        # Counting leading number of zeros above results in one too many bits
```
What is it about the implementation in NumPy that forces this slightly contrived behaviour? It'd be useful to add some comments here about the reasons behind it.

mpetroff commented:
I wrote the special handling of power of two ranges to fix a direct issue I found power of two ranges and thus specifically checked for such numbers. On a more careful examination of the CPython and NumPy random implementations to determine why they differed with power of two ranges, I came across [a comment](https://github.com/python/cpython/blob/v3.7.0/Lib/random.py#L233) that explains why the CPython implementation differs (to eliminate a special case) and makes for a much simpler fix&mdash;subtracting one from `n` before calculating the number of bits. Thus, I simplified the behavior (and added more comments).

stuartarchibald commented:
@mpetroff thanks for the fixes. It was definitely worth visiting the implementations to find the underlying reason behind the differences, thanks for taking the time to do that, the code is much cleaner as a result. I think this can be merged after shipping 0.39.0 is complete.
